File: AgenticOS/apps/web/lib/server/tavily.ts
import { z } from "zod";

const tavilyResultSchema = z.object({
  title: z.string().optional(),
  url: z.string().url().optional(),
  content: z.string().optional(),
  score: z.number().optional()
});

const tavilyResponseSchema = z.object({
  answer: z.string().optional(),
  results: z.array(tavilyResultSchema).optional()
});

export type TavilySource = {
  title: string;
  url: string;
  content: string;
  score?: number;
};

export async function searchWithTavily(query: string) {
  const apiKey = process.env.TAVILY_API_KEY;

  if (!apiKey) {
    return {
      answer: "",
      sources: [] as TavilySource[],
      error: "TAVILY_API_KEY is not configured."
    };
  }

  const response = await fetch("https://api.tavily.com/search", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json"
    },
    body: JSON.stringify({
      query,
      search_depth: "basic",
      include_answer: true,
      max_results: 5
    })
  });

  if (!response.ok) {
    return {
      answer: "",
      sources: [] as TavilySource[],
      error: `Tavily search failed with ${response.status}.`
    };
  }

  const parsed = tavilyResponseSchema.parse(await response.json());
  const sources =
    parsed.results
      ?.filter((result) => result.title && result.url)
      .map((result) => ({
        title: result.title!,
        url: result.url!,
        content: result.content ?? "",
        score: result.score
      })) ?? [];

  return {
    answer: parsed.answer ?? "",
    sources,
    error: null
  };
}

File: AgenticOS/apps/agent-api/app/services/tavily.py
import os

import httpx

from app.core.schemas import Source


async def search_tavily(query: str, api_key: str | None = None) -> tuple[str, list[Source], str | None]:
    api_key = api_key or os.getenv("TAVILY_API_KEY")
    if not api_key:
        return "", [], "TAVILY_API_KEY is not set. Add it to Railway environment variables or store it as a workspace API key."

    try:
        async with httpx.AsyncClient(timeout=20) as client:
            response = await client.post(
                "https://api.tavily.com/search",
                headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
                json={
                    "query": query,
                    "search_depth": "basic",
                    "include_answer": True,
                    "include_raw_content": False,
                    "max_results": 5,
                },
            )
            response.raise_for_status()
            payload = response.json()
    except Exception as exc:
        return "", [], f"Tavily search failed: {exc}"

    sources = [
        Source(
            title=str(item.get("title") or "Untitled source"),
            url=str(item.get("url") or ""),
            content=str(item.get("content") or ""),
            score=item.get("score"),
        )
        for item in payload.get("results", [])
        if item.get("url")
    ]

    return str(payload.get("answer") or ""), sources, None
